### PR TITLE
Feature/seeded bsk

### PR DESCRIPTION
**Pull request by IceTDrinker**
### Resolves

closes https://github.com/zama-ai/concrete-core-internal/issues/265

### Description

Seeded BSK, complete with C FFI

Missing formal defs

Test that checks the seeded versions are bit identical to non seeded once expand is added

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (apply the `aws_test` to the PR to launch the tests on AWS)
* [x] The draft release description has been updated
* [x] Check for breaking changes and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
aws tests : https://github.com/zama-ai/concrete-core/actions/runs/2576073845

for some reason they don't seem to appear in the checks

**Comment by IceTDrinker** [comment]
relaunching AWS which failed on the csprng uniformity test

File: concrete-core/src/commons/crypto/bootstrap/seeded_standard/mod.rs [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+    where
+        Cont: AsRefSlice<Element = Coef>,
+        Coef: UnsignedTorus,
+    {
+        let tensor = Tensor::from_container(cont);
+        ck_dim_div!(tensor.len() =>
+            decomp_level.0,
+            glwe_size.0,
+            poly_size.0
+        );
```
Comment (aPere3) [inline concrete-core/src/commons/crypto/bootstrap/seeded_standard/mod.rs:148]: also 2 ?

File: concrete-core/src/commons/crypto/bootstrap/mod.rs [hunk 2]
```diff
@@ -84,6 +89,100 @@ mod test {
...
+            par_seeded_bsk.par_fill_with_new_key::<_, _, _, _, _, SoftwareRandomGenerator>(
+                &lwe_sk,
+                &glwe_sk,
+                StandardDev::from_standard_dev(10.),
+                &mut DeterministicSeeder::<SoftwareRandomGenerator>::new(deterministic_seeder_seed),
+            );
+
+            par_seeded_bsk.expand_into::<_, _, SoftwareRandomGenerator>(&mut expanded_bsk);
+
+            assert_eq!(mono_bsk, expanded_bsk);
```
Comment (aPere3) [inline concrete-core/src/commons/crypto/bootstrap/mod.rs:182]: Is there a reason to have two tests in the same function ? Parallel and not parallel ? Whereas we have separate ones in the ggsw ?

File: concrete-core/src/backends/default/implementation/engines/default_parallel_engine/mod.rs [hunk 3]
```diff
@@ -36,7 +36,16 @@ type ActivatedRandomGenerator = AesniRandomGenerator;
...
+    /// A structure containing two CSPRNGs to generate material for encryption like public masks
+    /// and secret errors.
+    ///
+    /// The [`ImplEncryptionRandomGenerator`] contains two CSPRNGs, one publicly seeded used to
+    /// generate mask coefficients and one privately seeded used to generate errors during
+    /// encryption.
     pub(crate) encryption_generator: ImplEncryptionRandomGenerator<ActivatedRandomGenerator>,
+    /// A seeder that can be called to generate 128 bits seeds, useful to create new
+    /// [`ImplEncryptionRandomGenerator`] to encrypt seeded types.
+    seeder: ImplDeterministicSeeder<ActivatedRandomGenerator>,
```
Comment (aPere3) [inline concrete-core/src/backends/default/implementation/engines/default_parallel_engine/mod.rs:48]: seems funny that one field is `pub(crate)` and not the other. Don't know why we have that though :/

File: concrete-core/src/specification/engines/lwe_seeded_bootstrap_key_to_lwe_bootstrap_key_transformation.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+///
+/// ## LWE seeded bootstrap key to LWE bootstrap key transformation
+///
+/// TODO
+pub trait LweSeededBootstrapKeyToLweBootstrapKeyTransformationEngine<
+    InputSeededBootstrapKey,
+    OutputBootstrapKey,
+>: AbstractEngine where
+    InputSeededBootstrapKey: LweSeededBootstrapKeyEntity,
+    OutputBootstrapKey: LweBootstrapKeyEntity,
```
Comment (aPere3) [inline concrete-core/src/specification/engines/lwe_seeded_bootstrap_key_to_lwe_bootstrap_key_transformation.rs:28]: Just noting that there should also be bound here. But I guess you already had that in mind.

Comment (IceTDrinker) [reply concrete-core/src/specification/engines/lwe_seeded_bootstrap_key_to_lwe_bootstrap_key_transformation.rs:28]: the patch is waiting for the tests yes 🙂

File: concrete-core/src/backends/default/implementation/engines/default_parallel_engine/mod.rs (hunk 3 above)
Comment (IceTDrinker) [reply concrete-core/src/backends/default/implementation/engines/default_parallel_engine/mod.rs:48]: I did the chameleon here

File: concrete-core/src/commons/crypto/bootstrap/seeded_standard/mod.rs (hunk 1 above)
Comment (IceTDrinker) [reply concrete-core/src/commons/crypto/bootstrap/seeded_standard/mod.rs:148]: correct !

File: concrete-core/src/commons/crypto/bootstrap/mod.rs (hunk 2 above)
Comment (IceTDrinker) [reply concrete-core/src/commons/crypto/bootstrap/mod.rs:182]: Not quite sure now, I'll split

**Review by aPere3 — APPROVED** [review]

**Pull request merged**